tests/rsa_test.c

```diff
@@ -0,0 +1,691 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <aws/cal/cal.h>
+#include <aws/cal/hash.h>
+#include <aws/cal/private/der.h>
+#include <aws/cal/private/rsa.h>
+#include <aws/cal/rsa.h>
+#include <aws/common/encoding.h>
+#include <aws/testing/aws_test_harness.h>
+
+/*
+ * TODO: Need better test vectors. NIST ones are a pain to use.
+ * For now using manually generated vectors and relying on round tripping.
+ */
+
+static int s_byte_buf_decoded_from_base64_cur(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor cur,
+    struct aws_byte_buf *out) {
+    size_t decoded_length = 0;
+    ASSERT_SUCCESS(aws_base64_compute_decoded_len(&cur, &decoded_length));
+    ASSERT_SUCCESS(aws_byte_buf_init(out, allocator, decoded_length));
+    ASSERT_SUCCESS(aws_base64_decode(&cur, out));
+    return AWS_OP_SUCCESS;
+}
+
+static const char *TEST_ENCRYPTION_STRING = "The quick brown fox jumps over the lazy dog.";
+
+static int s_rsa_encryption_roundtrip_helper(
+    struct aws_allocator *allocator,
+    struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_encryption_algorithm algo) {
+    struct aws_byte_cursor plaintext_cur = aws_byte_cursor_from_c_str(TEST_ENCRYPTION_STRING);
+
+    /*short buffer should fail*/
+    struct aws_byte_buf ciphertext_short;
+    ASSERT_SUCCESS(aws_byte_buf_init(&ciphertext_short, allocator, 5));
+    ASSERT_ERROR(AWS_ERROR_SHORT_BUFFER, aws_rsa_key_pair_encrypt(key_pair, algo, plaintext_cur, &ciphertext_short));
+
+    /*make sure not to clobber anything in existing buffer*/
+    struct aws_byte_cursor prefix = aws_byte_cursor_from_c_str("random_prefix");
+    struct aws_byte_buf ciphertext;
+    ASSERT_SUCCESS(aws_byte_buf_init(&ciphertext, allocator, prefix.len + aws_rsa_key_pair_block_length(key_pair)));
```

Comment:
Is ciphertext (and your other output buffers) exactly the right size for the output you're writing, or do you also test the case where the output buffer is larger than necessary?

Reply:
the tests cover the buffer too small and buffer is block size cases (which for a lot of cases will probably be buffer is bigger than needed, once you strip off padding).